In [ ]:
import pandas as pd

In [1]:
import pandas as pd
titles = pd.read_csv(r'D:/Lets Upgrade AI-ML/Study material/titles.csv.csv')
cast = pd.read_csv(r'D:/Lets Upgrade AI-ML/Study material/cast.csv.csv')
titles.head()

title  year
0    Tasveer Mere Sanam  1996
1              Only You  1994
2  El pueblo del terror  1970
3               Machine  2007
4                  MARy  2008

### How many movies are listed in the titles dataframe?

In [2]:
titles.shape[0]

244914

#### 244914 movies are listed in the titles dataframe.

### What are the earliest two films listed in the titles dataframe?

In [3]:
df = titles.sort_values (by ='year')
df.head(2)

title  year
193222          Miss Jerry  1894
51388   The Startled Lover  1898

Miss Jerry and The Startled Lover are the two films listed earliest in the titles dataframe

### How many movies have the title "Hamlet"?

In [4]:
titles[titles['title'] == "Hamlet"].count()

title    20
year     20
dtype: int64

20 movies are titled as 'Hamlet'

### How many movies are titled "North by Northwest"?

In [5]:
titles[titles['title'] == "North by Northwest"].count()

title    1
year     1
dtype: int64

1 movie are titled 'North by Northwest'

### When was the first movie titled "Hamlet" made?

In [6]:
titles[titles['title'] == "Hamlet"].sort_values(by='year').head(1)

title  year
52521  Hamlet  1910

The first movie titled 'Hamlet' was made in 1910

### List all of the "Treasure Island" movies from earliest to most recent.

In [7]:
list1 = titles[titles['title'] == "Treasure Island"].sort_values (by='year').values.tolist()
print('List all of the "Treasure Island" movies from earliest to most recent is:')
List = pd.DataFrame(list1, columns =['Title', 'Year'])
print(List)

List all of the "Treasure Island" movies from earliest to most recent is:
             Title  Year
0  Treasure Island  1918
1  Treasure Island  1920
2  Treasure Island  1934
3  Treasure Island  1950
4  Treasure Island  1972
5  Treasure Island  1973
6  Treasure Island  1985
7  Treasure Island  1999


### How many movies were made in the year 1950?

In [8]:
titles[titles['year'] == 1950]['title'].count()

1113

1113 movies were made in the year 1950

### How many movies were made in the year 1960?

In [9]:
titles[titles['year'] == 1960]['title'].count()

1549

1549 movies were made in the year 1960

### How many movies were made from 1950 through 1959?

In [10]:
titles[(titles['year']>=1950)&(titles['year']<= 1959)]['title'].count()

12934

12934 movies were made from 1950 through 1959.

### In what years has a movie titled "Batman" been released?

In [14]:
list1=titles[titles['title'] =="Batman"]['year'].tolist()
print("The movie titled 'Batman' has been released in the years", list1)


The movie titled 'Batman' has been released in the years [1943, 1989]


### How many roles were there in the movie "Inception"?

In [15]:
cast[cast['title'] == 'Inception']['type'].count()

78

There are 78 roles in the movie inception 

### How many roles in the movie "Inception" are NOT ranked by an "n" value?

In [16]:
cast[cast['title'] == 'Inception']['n'].isnull().sum()

27

27 Roles in the movie 'Inception' are not ranked by an "n" value.

### But how many roles in the movie "Inception" did receive an "n" value?

In [17]:
cast[cast['title']=='Inception']['n'].notnull().sum()

51

51 Roles in the movie "Inception" Received an "n" value. 

### Display the cast of "North by Northwest" in their correct "n"-value order, ignoring roles that did not earn a numeric "n" value.

In [18]:
cast = cast[cast.n.notnull()].sort_values(by=['n'])
cast1 = cast[cast['title'] == 'North by Northwest'][['n','name']].values.tolist()
print("The cast of 'North by Northwest' in their correct 'n'-value order,ignoring roles that did not earn a numeric 'n' value are-")
Result = pd.DataFrame(cast1, columns = ['name', 'n']) 
print(Result)

The cast of 'North by Northwest' in their correct 'n'-value order,ignoring roles that did not earn a numeric 'n' value are-
    name                     n
0    1.0            Cary Grant
1    2.0       Eva Marie Saint
2    3.0           James Mason
3    4.0   Jessie Royce Landis
4    5.0        Leo G. Carroll
5    6.0  Josephine Hutchinson
6    7.0           Philip Ober
7    8.0         Martin Landau
8    9.0         Adam Williams
9   10.0          Edward Platt
10  11.0     Robert Ellenstein
11  12.0          Les Tremayne
12  13.0       Philip Coolidge
13  14.0         Patrick McVey
14  15.0          Edward Binns
15  16.0             Ken Lynch


### Display the entire cast, in "n"-order, of the 1972 film "Sleuth".

In [19]:
cast1=cast[cast.n.notnull()].sort_values(by=['n'])
cast1=cast1.loc[(cast1['title']=='Sleuth') & (cast1['year']==1972)][['name','n']]
print(cast1)

                       name    n
1705394    Laurence Olivier  1.0
326773        Michael Caine  2.0
373876       Alec Cawthorne  3.0
1465483  John (II) Matthews  4.0
2724553  Eve (III) Channing  5.0
1448690        Teddy Martin  6.0


The entire cast, in "n"-order,of the 1972 film "Sleuth" is^

### Now display the entire cast, in "n"-order, of the 2007 version of "Sleuth".

In [20]:
cast1=cast[cast.n.notnull()].sort_values(by=['n'])
cast1=cast1.loc[(cast1['title']=='Sleuth') & (cast1['year']==2007)][['name','n']]
print(cast1)

                  name    n
326774   Michael Caine  1.0
1293634       Jude Law  2.0
1805907  Harold Pinter  3.0


 The entire cast, in "n"-order, of the 2007 version of "Sleuth" 

### How many roles were credited in the silent 1921 version of Hamlet?

In [21]:
cast1= cast.loc[(cast['title']=='Hamlet')&(cast['year']==1921)]
cast1['character'].count()

9

9 Roles were credited in the silent 1921 version of Hamlet

### How many roles were credited in Branagh’s 1996 Hamlet?

In [22]:
cast2= cast.loc[(cast['title']=='Hamlet')&(cast['year']==1996)]
cast2['character'].count()

46

46 roles were credited in Branagh’s 1996 Hamlet.

### How many "Hamlet" roles have been listed in all film credits through history?

In [23]:
cast1 = cast[cast['title'] == 'Hamlet']
cast1['character'].count()

249

249 "Hamlet" roles have been listed in all film credits through history.

### How many people have played an "Ophelia"?

In [24]:
cast[cast['character']=='Ophelia']['name'].count()

60

60 People have played in "Ophelia"

### How many people have played a role called "The Dude"?

In [25]:
cast[cast['character']=='The Dude']['name'].count()

12

12  people have played a role called "The Dude"

### How many people have played a role called "The Stranger"?

In [26]:
cast[cast['character']=='The Stranger']['name'].count()

133

133 people have played a role called "The Stranger"

### How many roles has Sidney Poitier played throughout his career?

In [27]:
cast[cast['name']=='Sidney Poitier']['character'].count()

42

Sidney Poitier has played 42 roles throughout his career.

### How many roles has Judi Dench played?

In [28]:
cast[cast['name']=='Judi Dench']['character'].count()

51

Judi Dench has played 51 roles throughout his career.

### List the supporting roles (having n=2) played by Cary Grant in the 1940s, in order by year.

In [29]:
cast1=cast[(cast['name']=="Cary Grant")&(cast['n']==2)&(cast['year']>=1940)&(cast['year']<1950)]
list1 = cast1[['character','year']].values.tolist()
Result = pd.DataFrame(list1, columns = ['Supporting Role', 'Year'])
print(Result)

  Supporting Role  Year
0     Roger Adams  1941
1      Nick Arden  1940


^ These are the supporting roles played by Cary Grant in the 1940s.

### List the leading roles that Cary Grant played in the 1940s in order by year.

In [30]:
cast1 = cast[(cast['name']=="Cary Grant")&(cast['n']==1)&(cast['year']>=1940)&(cast['year']< 1950)].sort_values(by = ['year'])
list1= cast1[['character','year']].values.tolist()
result1 = pd.DataFrame(list1, columns = ['Leading Role', 'Year'])
print(result1)

                            Leading Role  Year
0                           Walter Burns  1940
1                     C. K. Dexter Haven  1940
2                            Matt Howard  1940
3                       Johnnie Aysgarth  1941
4                  Patrick 'Pat' O'Toole  1942
5                           Leopold Dilg  1942
6                          Capt. Cassidy  1943
7   Joe Adams -posing as Joe Bascopolous  1943
8                            Jerry Flynn  1944
9                             Ernie Mott  1944
10                     Mortimer Brewster  1944
11                                Devlin  1946
12                           Cole Porter  1946
13                                Dudley  1947
14                           Dick Nugent  1947
15                     Dr. Madison Brown  1948
16                         Jim Blandings  1948
17                 Captain Henri Rochard  1949


^ These are the leading roles played by Cary Grant in the 1940s.

### How many roles were available for actors in the 1950s?

In [31]:
cast[(cast['year']>=1950)&(cast['year']<1960)&
     (cast['type']=='actor')]['character'].count()

82079

82079 roles were available for actors in the 1950's.

### How many roles were available for actresses in the 1950s?

In [32]:
cast[(cast['year']>=1950)&(cast['year']<1960)&
     (cast['type']=='actress')]['character'].count()

35794

35794 roles were available for actors in 1950's.

### How many leading roles (n=1) were available from the beginning of film history through 1980?

In [33]:
cast[(cast['year']>=1894)&(cast['year']<= 1980)&
     (cast['n']==1)]['character'].count()

65140

65140 leading roles (n=1) were available from the beginning of film history through 1980. 

### How many non-leading roles were available through from the beginning of film history through 1980?

In [34]:
cast[(cast['year']>=1894)&(cast['year']<=1980)&
     (cast['n'])]['character'].count()

734460

734460 non-leading roles were available through from the beginning of film history through 1980.

### How many roles through 1980 were minor enough that they did not warrant a numeric "n" rank?

In [ ]:
cast[(cast['year'] >= 1970) & (cast['year'] <= 1980)]['n'].isnull().sum()

448347 roles through 1980 were minor enough that they did not warrant a numeric "n" rank.